blockchain/scriptval.go

```diff
@@ -227,6 +257,25 @@ func checkBlockScripts(block *btcutil.Block, utxoView *UtxoViewpoint, scriptFlag
 	}
 	txValItems := make([]*txValidateItem, 0, numInputs)
 	for _, tx := range block.Transactions() {
+		sha := tx.Hash()
+
+		// If the HashCache is present, and it doesn't yet contain the
+		// partial sighashes for this transaction, then we add the
+		// sighashes for the transaction. This allows us to take
+		// advantage of the potential speed savings due to the new
+		// digest algorithm (BIP0143).
+		// TODO(roasbeef): guard with segwit script flag
+		if hashCache != nil && !hashCache.ContainsHashes(sha) {
+			hashCache.AddSigHashes(tx.MsgTx())
+		}
+
+		var cachedHashes *txscript.TxSigHashes
```

Comment:
I see what you're saying here, but the way the the hash caching is written in this manner was due to an optimization I added when I was creating mega blocks on testnet way back last year. By checking and populating the hash cache here, I'm able to take advantage of the fact that the `txid`'s of each transaction should already be cached within the `btcutil.Tx` struct. When profiling those large blocks (which had transactions with thouands of inputs and outputs), I found a significant amount of time was spent just re-calculating the `txid` each time to order to check the cache for existence of the sighashes. 

The downside of this approach compared to pushing the hash cache all the way down the stack is that this initialization is done serially rather than concurrently within each of the validators on a need basis.